gpytorch/priors/prior.py

```diff
@@ -1,69 +1,18 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-from __future__ import unicode_literals
-
-from abc import abstractmethod, abstractproperty
-import torch
-from torch.nn import Module
+from __future__ import absolute_import, division, print_function, unicode_literals
 
+from abc import ABC
 
-class Prior(Module):
-    """
-    Base class for Priors in GPyTorch.
-    In GPyTorch, a parameter can be assigned a prior by passing it as the `prior` argument to
-    :func:`~gpytorch.module.register_parameter`. GPyTorch performs internal bookkeeping of priors,
-    and for each parameter with a registered prior includes the log probability of the parameter under its
-    respective prior in computing the Marginal Log-Likelihood (see e.g. :func:`~gpytorch.priors.Prior.forward`).
-
-    In order to define a new Prior in GPyTorch, a user must define at a minimum the following methods
-    * :func:`~gpytorch.priors.Prior.size`, which returns the size of the domain as a torch.Size.
-    * :func:`~gpytorch.priors.Prior.is_in_support`, which for a given parameter value returns a bool indicating
-        whether the parameter is contained in the support of the distribution.
-    * :func:`~gpytorch.priors.Prior._log_prob`, which returns the log-probability for a given parameter value
-        as a torch scalar.
+from torch.distributions import Distribution
+from torch.nn import Module
 
-    """
 
-    @abstractmethod
-    def is_in_support(self, parameter):
-        raise NotImplementedError()
+class Prior(Distribution, Module, ABC):
+    """Priors ar imposed on hyperparameters"""
```

Comment:
Just a thought (though this is probably for a later refactor): we're not really using any functionality from Module here. As far as I can tell, the only thing we're getting from it right now is the `cpu` and `cuda` functionality.

I think what might be best as a future move: we should create thin wrappers around all the distributions that we care about, and add `cpu`/`cuda`/`_apply` methods. I imagine that this could be something that could be eventually merged upstream into PyTorch.

This is a post beta-release refactor I think, but planting the idea now.

Reply:
Yeah so the other thing that's kind of nice about these being modules is that when working interactively and print your top-level module you see all the priors attached as well. 

I remember making the suggestion of adding `cpu` and `cuda` methods upstream in pytorch, but that wasn't all that popular (it wasn't that it was unpopular, people just thought it's not needed, the rationale being that Distributions shouldn't be persisting and you can always manually re-instantiate a cuda version). But I think it's something we'll want otherwise things get ugly.